ddtrace/tracer/spancontext.go

```diff
@@ -414,10 +415,82 @@ func (t *trace) finishedOne(s *span) {
 	if hn := tr.hostname(); hn != "" {
 		s.setMeta(keyTracerHostname, hn)
 	}
+	setPeerService(s, tr.config)
 	// we have a tracer that can receive completed traces.
 	atomic.AddUint32(&tr.spansFinished, uint32(len(t.spans)))
 	tr.pushTrace(&finishedTrace{
 		spans:    t.spans,
 		willSend: decisionKeep == samplingDecision(atomic.LoadUint32((*uint32)(&t.samplingDecision))),
 	})
 }
+
+// setPeerService sets the peer.service, _dd.peer.service.source, and _dd.peer.service.remapped_from
+// tags as applicable for the given span.
+func setPeerService(s *span, cfg *config) {
+	if _, ok := s.Meta[ext.PeerService]; ok { // peer.service already set on the span
+		s.setMeta(keyPeerServiceSource, ext.PeerService)
+	} else { // no peer.service currently set
+		spanKind := s.Meta[ext.SpanKind]
+		isOutboundRequest := spanKind == ext.SpanKindClient || spanKind == ext.SpanKindProducer
+		shouldSetDefaultPeerService := isOutboundRequest && cfg.peerServiceDefaultsEnabled
+		if !shouldSetDefaultPeerService {
+			return
+		}
+		source := setPeerServiceFromSource(s)
+		if source == "" {
+			log.Warn("No source tag value could be found for span %q, peer.service not set", s.Name)
+			return
+		}
+		s.setMeta(keyPeerServiceSource, source)
+	}
+	// Overwrite existing peer.service value if remapped by the user
+	ps := s.Meta[ext.PeerService]
+	if to, ok := cfg.peerServiceMappings[ps]; ok {
+		s.setMeta(keyPeerServiceRemappedFrom, ps)
+		s.setMeta(ext.PeerService, to)
+	}
+}
+
+// setPeerServiceFromSource sets peer.service from the sources determined
+// by the tags on the span. It returns the source tag name that it used for
+// the peer.service value, or the empty string if no valid source tag was available.
+func setPeerServiceFromSource(s *span) string {
+	has := func(tag string) bool {
+		_, ok := s.Meta[tag]
+		return ok
+	}
+	var sources []string
+	switch {
+	case has(ext.DBSystem):
+		sources = []string{
+			ext.CassandraContactPoints,
+			ext.DBInstance,
+			ext.DBName,
+			ext.PeerHostname,
+			ext.TargetHost}
+	case has(ext.MessagingSystem):
+		sources = []string{
+			ext.KafkaBootstrapServers,
+			ext.PeerHostname,
+			ext.TargetHost,
+		}
+	case has(ext.RPCSystem):
+		sources = []string{
+			ext.RPCService,
+			ext.PeerHostname,
+			ext.TargetHost,
+		}
+	default:
+		sources = []string{
+			ext.PeerHostname,
+			ext.TargetHost,
+		}
+	}
```

Comment:
I might be missing something obvious, but can't we just put all the potential sources in one list and iterate over it? I don't understand why we need the switch cases 🤔

Reply:
I had a few things in mind when writing this code, and the first one was trying to reflect the algorithm the same as in the RFC (if it's an AWS span, do this, else if it is a DB span, do that, etc.)

Also, considering this code runs for every span (including for manual instrumentation and custom tags, not only our contribs), I wanted to ensure the existence of some basic tags we set in our integrations (like `db.system`, etc.) before using those tags as sources/precursors.

In addition, I recently updated the PR adding the AWS ones, which sadly are unprefixed (`tablename`, `bucketname`, etc.), and I want to make sure of at least the existence of `aws_service` before using those tags as sources.

Not very strongly opinionated about this, as technically it could also work having a single list with all the precursors, but if you have any concerns about readability/performance I am happy to change it.

Reply:
I understand this can be very subjective, but I personally don't see how the code is more readable this way. [One of the Go idioms is to keep things simple](https://dave.cheney.net/2020/02/23/the-zen-of-go).

Reply:
Discussed offline, and I agreed with @pkalmakis and @rarguelloF that this will work fine for the time being. When we explored the other approach, it would have different behavior and the readability would be worse. They can go and refactor this later if needed.

Reply:
Talked about this more, it's also important since some of these tags are only relevant in certain situations and we DONT want to use them as a source for peer.service outside those situations (e.g. `topicname` might be set on a kafka stream span but we DONT want to use it for the peer.service and instead want to use the bootstrap server and other tags)